File: AndroidServiceCommunication/app/src/main/java/app/serviceconnector/activities/ServiceLauncherActivity.java
package app.serviceconnector.activities;

import android.app.Activity;
import android.os.Bundle;
import android.os.Message;
import android.view.View;
import android.widget.Toast;

import app.serviceconnector.R;
import lib.serviceconnector.ServiceManager;

public final class ServiceLauncherActivity extends Activity implements View.OnClickListener {

    private ServiceManager mServiceManager;
    private Class service;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_demo);

        mServiceManager = new ServiceManager(this, new ServiceManager.MessageReceiver() {
            @Override
            public void onMessageReceivedFromService(Object object) {
                if (object != null) {
                    Toast.makeText(getApplicationContext(), object + "", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onMessageReceivedFromService(Message message) {
                if (message != null) {
                    Toast.makeText(getApplicationContext(), message + "", Toast.LENGTH_SHORT).show();
                }
            }
        });
        findViewById(R.id.btn_send).setOnClickListener(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        Bundle bundle = getIntent().getExtras();
        if (bundle != null) {
            String serviceName = bundle.getString("SERVICE");
            try {
                service = Class.forName(serviceName);
                mServiceManager.startService(service);
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onClick(View v) {
        mServiceManager.sendMessageToService("Message sent from UI to " + service.getSimpleName());
    }

    @Override
    protected void onDestroy() {
        mServiceManager.disconnectedToService();
        super.onDestroy();
    }
}

File: Activity-Service-Communication/app/src/main/java/app/serviceconnector/activities/ServiceLauncherActivity.kt
package app.serviceconnector.activities

import android.app.Activity
import android.os.Bundle
import android.os.Message
import android.view.View
import android.widget.Toast

import app.serviceconnector.R
import kotlinx.android.synthetic.main.activity_demo.*
import lib.connector.MessageReceiver
import lib.connector.ServiceManager

class ServiceLauncherActivity : Activity(), View.OnClickListener {

    private var serviceManager: ServiceManager? = null
    private var service: Class<*>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_demo)

        serviceManager = ServiceManager(this, object : MessageReceiver {

            override fun onMessageReceivedFromService(anyObject: Any?) {
                anyObject?.let {
                    Toast.makeText(applicationContext, it.toString() + "", Toast.LENGTH_SHORT)
                        .show()
                }
            }

            override fun onMessageReceivedFromService(message: Message?) {
                message?.let {
                    Toast.makeText(applicationContext, message.toString() + "", Toast.LENGTH_SHORT)
                        .show()
                }
            }
        })
        sendBtn.setOnClickListener(this)
        stopBtn.setOnClickListener(this)
    }

    override fun onStart() {
        super.onStart()
        val bundle = intent.extras
        if (bundle != null) {
            val serviceName = bundle.getString("SERVICE")
            try {
                service = Class.forName(serviceName!!)
                serviceManager?.startService(service!!)
            } catch (e: ClassNotFoundException) {
                e.printStackTrace()
            }

        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.sendBtn -> {
                serviceManager?.sendMessageToService("Message sent from UI to " + service!!.simpleName)
            }
            R.id.stopBtn -> {
                serviceManager?.stopService()
            }
        }
    }

    override fun onDestroy() {
        serviceManager?.disconnectedToService()
        super.onDestroy()
    }
}
